src/sqlfluff/core/rules/crawlers.py

```diff
@@ -0,0 +1,125 @@
+"""Definitions of crawlers."""
+
+from typing import Iterator, Set, cast
+from sqlfluff.core.parser.segments.base import BaseSegment
+from sqlfluff.core.parser.segments.raw import RawSegment
+
+from sqlfluff.core.rules.context import RuleContext
+
+
+class BaseCrawler:
```

Comment:
We could declare this class as an abstract base class to prevent creating it. I think that'd remove the need for the special `crawl()` implementation that just raises an error.

https://docs.python.org/3/library/abc.html